Linux port via Blade

kvark opened the pull request:
## Motivation

I :heart:  Zed! It's lightning fast and has great UX. I want it to run as well on all major platforms. I'm currently using Linux most actively.
[Blade](https://github.com/kvark/blade) is a good candidate for providing GPU access: it supports Vulkan, Metal, and GLES/WebGL. Its abstraction is extremely thin, while having one of the nicest GPU APIs. Codebase is also tiny. Checkout [the meetup recording](https://www.youtube.com/watch?v=63dnzjw4azI&t=623s) from a year ago.
I believe these projects make a good match :rocket: !

### Why this is a bad idea

If Zed team wants to use off-the-shelf components from Rust ecosystem, then Blade is certainly at disadvantage here, since it's not widely used. It would rely on Zed team adding necessary features in a branch, then maybe upstreaming some of them. That is to say, it's unclear if this can be avoided with more popular alternatives - being flexible with any local changes is a good ability.

### Why it's not too bad

Blade uses [WGSL](https://www.w3.org/TR/WGSL) shaders, similar to `wgpu` and `arcana`, but without the binding decorations. So this aspect of the product is nicely portable.

## Progress

- [ ] Platforms
  - [x] X11 (via xcb)
    - [ ] input handling
    - [ ] get proper content size
  - [ ] Windows
  - [ ] Replace the existing Metal backend
- [ ] Text System
  - [ ] shaping
  - [ ] glyph rasterization
- [x] Texture atlas
- [ ] Rendering
  - [x] basic primitives
  - [x] path rendering
  - [x] sprite rendering
- [ ] media surfaces
- [ ] CI

## Current status
Zed starts up but crashes on text-system related checks.
![zed-linux-1](https://github.com/zed-industries/zed/assets/107301/ba536218-4d2c-43c9-ae6c-bef69b54bd0c)

kvark committed:
Add blade dependency

kvark committed:
Add Linux platform, gate usage of CVImageBuffer by macOS

kvark committed:
linux: implement dispatcher, add dummy textsystem

kvark committed:
linux: start the text system

kvark committed:
linux: basic window, display, and atlas

kvark committed:
linux: hook up X11rb for Window creation

kvark committed:
x11: create window and route events

kvark committed:
linux: port from x11rb to xcb and hook up RawWindowHandle

kvark committed:
linux: hook up render event, basic renderer command buffer

kvark committed:
linux: work around the mutex locks for request_frame and resize

kvark committed:
linux: basic quad renderer logic

kvark committed:
hide MacOS dependencie of live_kit_client and media

kvark committed:
linux: shadow rendering

kvark committed:
linux: share corner picking code between shaders

kvark committed:
linux: refactor window structure, support move callback

kvark committed:
WIP path rasterization

kvark committed:
linux: implement RWH for LinuxWindow

kvark committed:
linux: path rasterization shader

kvark committed:
blade: path sprite rendering

kvark committed:
blade: refactor the atlas to not own the command encoder

kvark committed:
blade: port underline shader

kvark committed:
linux: various fixes across the crates to make it compile

kvark committed:
blade: point to master, remove the override

kvark commented:
@cla-bot check

kvark committed:
blade: mono/poly chrome sprite rendering

kvark committed:
fix MacOS build, switch external RWH to 0.6

leaving blade-internal RWH as 0.5 until this is fixed:
https://github.com/ash-rs/ash/issues/864

kvark committed:
blade: fix shadow vertex bounds

kvark committed:
blade: enforce alignment in the belt

istudyatuni commented on rust-toolchain.toml:
```diff
@@ -2,4 +2,4 @@
 profile = "minimal"
 components = [ "rustfmt", "clippy" ]
-targets = [ "x86_64-apple-darwin", "aarch64-apple-darwin", "wasm32-wasi" ]
+targets = [ "x86_64-apple-darwin", "aarch64-apple-darwin", "x86_64-linux-gnu", "wasm32-wasi" ]
```
```suggestion
targets = [ "x86_64-apple-darwin", "aarch64-apple-darwin", "x86_64-unknown-linux-gnu", "wasm32-wasi" ]
```

kvark committed:
blade: tune belt alignment to match Intel Iris Xe

kvark committed:
Fix linux target in rust-toolchain.toml

Co-authored-by: Ilia <43654815+istudyatuni@users.noreply.github.com>

kvark committed:
blade: always create texture views for atlas tiles

kvark committed:
blade: initialize atlas textures

kvark committed:
blade: encapsulate BladeAtlasStorage for indexing

williamdes commented:
I can confirm it builds on Debian 12. But I had to revert b7ced394 because the compiling never ends. (/cc @d1y do you know why ?)

For anyone needed this, here is my list of commands I kept in my notes to get zed to build:
```sh
sudo apt install libclang-dev

wget https://snapshot.debian.org/archive/debian/20221113T151655Z/pool/main/libd/libdispatch/libdispatch-dev_0~svn197-4_amd64.
wget https://snapshot.debian.org/archive/debian/20221113T151655Z/pool/main/libd/libdispatch/libdispatch0_0~svn197-4_amd64.deb
wget http://ftp.debian.org/debian/pool/main/libk/libkqueue/libkqueue-dev_2.3.1-1_amd64.deb
wget http://ftp.debian.org/debian/pool/main/libk/libkqueue/libkqueue0_2.3.1-1_amd64.deb
wget http://ftp.debian.org/debian/pool/main/libp/libpthread-workqueue/libpthread-workqueue0_0.9.1-1_amd64.deb
wget http://ftp.debian.org/debian/pool/main/libp/libpthread-workqueue/libpthread-workqueue-dev_0.9.1-1_amd64.deb

sudo apt install ./libdispatch-dev_0~svn197-4_amd64.deb ./libdispatch0_0~svn197-4_amd64.deb ./libkqueue-dev_2.3.1-1_amd64.deb ./libpthread-workqueue-dev_0.9.1-1_amd64.deb ./libkqueue0_2.3.1-1_amd64.deb ./libpthread-workqueue0_0.9.1-1_amd64.deb


sudo apt install  gobjc-12

# -> GPU
sudo apt install libfontconfig-dev
```

/cc @aminya

kvark committed:
blade: fix tile bounds shader FFI

williamdes commented:
```sh
sudo apt install libc6-dev
```

I applied this diff but still get a  segfault:
```diff
diff --git a/crates/collab/Cargo.toml b/crates/collab/Cargo.toml
index 0fb8cb929..1e1d0f916 100644
--- a/crates/collab/Cargo.toml
+++ b/crates/collab/Cargo.toml
@@ -37,7 +37,7 @@ parking_lot.workspace = true
 prometheus = "0.13"
 prost.workspace = true
 rand.workspace = true
-reqwest = { version = "0.11", features = ["json"], optional = true }
+reqwest = { version = "0.11", features = ["json", "rustls-tls"], optional = true, no-default-features = true }
 rpc = { path = "../rpc" }
 scrypt = "0.7"
 sea-orm = { version = "0.12.x", features = ["sqlx-postgres", "postgres-array", "runtime-tokio-rustls", "with-uuid"] }
diff --git a/crates/live_kit_server/Cargo.toml b/crates/live_kit_server/Cargo.toml
index 63b4fe106..a4b1334db 100644
--- a/crates/live_kit_server/Cargo.toml
+++ b/crates/live_kit_server/Cargo.toml
@@ -19,7 +19,7 @@ jwt = "0.16"
 log.workspace = true
 prost = "0.8"
 prost-types = "0.8"
-reqwest = "0.11"
+reqwest = { version = "0.11", features = ["rustls-tls"], no-default-features = true }
 serde.workspace = true
 serde_derive.workspace = true
 sha2 = "0.10"
```

Do you know why ?
How to you debug crashes ?

Here is my gdb output:
```
Reading symbols from ./target/debug/Zed...
(gdb) run
Starting program: /mnt/Dev/@williamdes/@forked-repos/zed/target/debug/Zed 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x0000000000000001 in ?? ()
(gdb) bt
#0  0x0000000000000001 in ?? ()
#1  0x00007fffffffe217 in ?? ()
#2  0x0000000000000000 in ?? ()
```

aminya commented on crates/cli/Cargo.toml:
```diff
@@ -23,6 +23,6 @@ serde_derive.workspace = true
 [target.'cfg(target_os = "macos")'.dependencies]
-core-foundation = "0.9"
-core-services = "0.2"
-plist = "1.3"
+#core-foundation = "0.9"
+#core-services = "0.2"
+#plist = "1.3"
```
This is already guarded with the MacOS target. It should not be needed to guard it again.

aminya commented on crates/gpui/Cargo.toml:
```diff
@@ -26,6 +26,9 @@ anyhow.workspace = true
 async-task = "4.7"
 backtrace = { version = "0.3", optional = true }
 bitflags = "2.4.0"
+blade-graphics = { git = "https://github.com/kvark/blade", rev = "62eb18d312f720a5aac8f508fe223146a24fc7f0" }
+blade-macros = { git = "https://github.com/kvark/blade", rev = "62eb18d312f720a5aac8f508fe223146a24fc7f0" }
+bytemuck = "1"
```
These added dependencies should be guarded for non-MacOS targets

kvark replied:
My plan is to use this path on all platforms. But I agree - for the sake of landing it sooner let's focus on Linux only.

aminya commented on crates/gpui/src/platform/linux/platform.rs:
```diff
@@ -0,0 +1,302 @@
+        unimplemented!()
+    }
+
+    fn reveal_path(&self, path: &Path) {}
+
+    fn on_become_active(&self, callback: Box<dyn FnMut()>) {}
+
+    fn on_resign_active(&self, callback: Box<dyn FnMut()>) {}
+
+    fn on_quit(&self, callback: Box<dyn FnMut()>) {}
+
+    fn on_reopen(&self, callback: Box<dyn FnMut()>) {}
+
+    fn on_event(&self, callback: Box<dyn FnMut(PlatformInput) -> bool>) {}
+
+    fn on_app_menu_action(&self, callback: Box<dyn FnMut(&dyn Action)>) {}
+
+    fn on_will_open_app_menu(&self, callback: Box<dyn FnMut()>) {}
+
+    fn on_validate_app_menu_command(&self, callback: Box<dyn FnMut(&dyn Action) -> bool>) {}
```
Should these have `unimplementated!()`?

aminya commented on crates/live_kit_client/Cargo.toml:
```diff
@@ -19,7 +19,7 @@ test-support = [
     "collections/test-support",
```
Live-kit was already compilable on Linux after #6859, and it is being worked on in #7302. Are these changes needed?

kvark replied:
Unsure, but let me know when would be the best time to pull latest from `main` into the PR branch, so that live kit doesn't need patches

aminya reviewed (changes requested):
It is so exciting to see this progressing! Thank you @kvark! 💯

aminya commented:
> For anyone needed this, here is my list of commands I kept in my notes to get zed to build:
> 
> ```shell
> sudo apt install libclang-dev
> 
> wget https://snapshot.debian.org/archive/debian/20221113T151655Z/pool/main/libd/libdispatch/libdispatch-dev_0~svn197-4_amd64.
> wget https://snapshot.debian.org/archive/debian/20221113T151655Z/pool/main/libd/libdispatch/libdispatch0_0~svn197-4_amd64.deb
> wget http://ftp.debian.org/debian/pool/main/libk/libkqueue/libkqueue-dev_2.3.1-1_amd64.deb
> wget http://ftp.debian.org/debian/pool/main/libk/libkqueue/libkqueue0_2.3.1-1_amd64.deb
> wget http://ftp.debian.org/debian/pool/main/libp/libpthread-workqueue/libpthread-workqueue0_0.9.1-1_amd64.deb
> wget http://ftp.debian.org/debian/pool/main/libp/libpthread-workqueue/libpthread-workqueue-dev_0.9.1-1_amd64.deb
> 
> sudo apt install ./libdispatch-dev_0~svn197-4_amd64.deb ./libdispatch0_0~svn197-4_amd64.deb ./libkqueue-dev_2.3.1-1_amd64.deb ./libpthread-workqueue-dev_0.9.1-1_amd64.deb ./libkqueue0_2.3.1-1_amd64.deb ./libpthread-workqueue0_0.9.1-1_amd64.deb
> 
> 
> sudo apt install  gobjc-12
> 
> # -> GPU
> sudo apt install libfontconfig-dev
> ```


If there are system dependencies, it should be added to this script so that they are installed during the bootstrap:
https://github.com/zed-industries/zed/blob/main/script/linux

kvark committed:
linux: disable mac-os specific build commands

kvark committed:
linux: temporarily disable purescript tree sitter

Link error:
  = note: /nix/store/idiaraknw071d20nlqp49s18gbvw4wa0-binutils-2.40/bin/ld: /x/Code/zed/target/x86_64-unknown-linux-gnu/debug/deps/libtree_sitter_haskell-7323f782ad886c6d.rlib(scanner.o): in function `state_new':
          /home/kvark/.cargo/git/checkouts/tree-sitter-haskell-74c278e7a2ef8d7d/cf98de2/src/scanner.c:218: multiple definition of `state_new'; /x/Code/zed/target/x86_64-unknown-linux-gnu/debug/deps/libtree_sitter_purescript-b0a95fb604a5817c.rlib(scanner.o):/home/kvark/.cargo/git/checkouts/tree-sitter-purescript-88dd3ec656c48026/a37140f/src/scanner.c:218: first defined here

kvark committed:
linux: only tick the main thread tasks and one at a time in the event loop

kvark committed:
linux: switch folders to use CONFIG_DIR

kvark committed:
linux: HACK disable watcher and crash uploader temporarily

kvark committed:
linux: refactor LinuxPlatform to avoid recursive locking from the window callbacks

aminya committed:
build: add Blade font-config sys deps on Linux

aminya commented on crates/fs/src/fs.rs:
```diff
@@ -301,6 +301,8 @@ impl Fs for RealFs {
             .configure(Config::default().with_poll_interval(latency))
             .unwrap();
 
+        //TODO: unblock this
+        #[cfg(not(target_os = "linux"))]
         watcher
             .watch(path, notify::RecursiveMode::Recursive)
             .unwrap();
```
This results in an empty user settings which results in an error during startup. Why was it disabled? 
```
user_settings_content: 
[2024-02-04T23:15:58-08:00 ERROR util] crates/settings/src/settings_file.rs:77: EOF while parsing a value at line 1 column 0
```

aminya replied:
Can be fixed via [kvark/zed@`daa2c84` (#1)](https://github.com/kvark/zed/pull/1/commits/daa2c8443dc4994700d021f43dea29bdeab8b20d)

kvark replied:
great stuff, thank you!

aminya committed:
fix: create the settings and keymaps to unblock file watching

aminya commented:
After https://github.com/kvark/zed/pull/1, the Zed CLI runs without an error, but I cannot see any UI.

kvark committed:
linux: create a hidden window inside the platform

It allows us to receive messages from the dispatcher,
which breaks us out of waiting and lets us execute
main thread runnables as a part of the main loop.

williamdes commented:
I now have this crash and a white window that shows 1 second. Merged  kvark/zed#2 locally.
This is awesome progress ! 🎉 

<details><summary>output</summary>
<p>

```
Thread "main" panicked with "called `Result::unwrap()` on an `Err` value: ERROR_LAYER_NOT_PRESENT" at /home/williamdes/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:281:55
   0: Zed::init_panic_hook::{closure#0}
             at crates/zed/src/main.rs:583:29
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:783:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:657:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:170:18
   4: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   5: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   6: core::result::unwrap_failed
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
   7: <core::result::Result<ash::instance::Instance, ash::vk::enums::Result>>::unwrap
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
   8: <blade_graphics::hal::Context>::init_impl
             at /home/williamdes/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:281:13
   9: <blade_graphics::hal::Context>::init_windowed::<gpui::platform::linux::window::RawWindow>
             at /home/williamdes/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:562:9
  10: <gpui::platform::linux::window::LinuxWindowState>::new
             at crates/gpui/src/platform/linux/window.rs:196:17
  11: <gpui::platform::linux::platform::LinuxPlatform as gpui::platform::Platform>::open_window
             at crates/gpui/src/platform/linux/platform.rs:193:35
  12: <gpui::window::Window>::new
             at crates/gpui/src/window.rs:332:31
  13: <gpui::app::AppContext>::open_window::<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>::{closure#0}
             at crates/gpui/src/app.rs:482:30
  14: <gpui::app::AppContext>::update::<gpui::window::WindowHandle<workspace::Workspace>, <gpui::app::AppContext>::open_window<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>::{closure#0}>
             at crates/gpui/src/app.rs:357:22
  15: <gpui::app::AppContext>::open_window::<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>
             at crates/gpui/src/app.rs:479:9
  16: <gpui::app::async_context::AsyncAppContext>::open_window::<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>
             at crates/gpui/src/app/async_context.rs:133:12
  17: <workspace::Workspace>::new_local::{closure#0}::{closure#0}
             at crates/workspace/src/workspace.rs:838:17
  18: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>>> as core::future::future::Future>::poll
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/future/future.rs:125:9
  19: <<async_task::runnable::Builder<_>>::spawn_local::Checked<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>>>> as core::future::future::Future>::poll
             at /home/williamdes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/runnable.rs:455:26
  20: <async_task::raw::RawTask<<async_task::runnable::Builder<_>>::spawn_local::Checked<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>>>>, core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>, <gpui::executor::ForegroundExecutor>::spawn::inner<core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>::{closure#0}, ()>>::run
             at /home/williamdes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/raw.rs:557:17
  21: <async_task::runnable::Runnable>::run
             at /home/williamdes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/runnable.rs:781:18
  22: <gpui::platform::linux::platform::LinuxPlatform as gpui::platform::Platform>::run
             at crates/gpui/src/platform/linux/platform.rs:144:17
  23: <gpui::app::App>::run::<Zed::main::{closure#3}>
             at crates/gpui/src/app.rs:138:9
  24: Zed::main
             at crates/zed/src/main.rs:124:5
  25: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
  26: std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:154:18
  27: std::rt::lang_start::<()>::{closure#0}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:167:18
  28: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
      std::panicking::try
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
      std::panic::catch_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:48
      std::panicking::try::do_call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
      std::panicking::try
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
      std::panic::catch_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:20
  29: std::rt::lang_start::<()>
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:166:17
  30: main
  31: <unknown>
  32: __libc_start_main
  33: _start


```

</p>
</details>

nhemsley commented:
> I now have this crash and a white window that shows 1 second. Merged [kvark#2](https://github.com/kvark/zed/pull/2) locally. This is awesome progress ! 🎉
> 
> output
> ```
> Thread "main" panicked with "called `Result::unwrap()` on an `Err` value: ERROR_LAYER_NOT_PRESENT" at /home/williamdes/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:281:55
>    0: Zed::init_panic_hook::{closure#0}
>              at crates/zed/src/main.rs:583:29
>    1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2021:9
>       std::panicking::rust_panic_with_hook
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:783:13
>    2: std::panicking::begin_panic_handler::{{closure}}
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:657:13
>    3: std::sys_common::backtrace::__rust_end_short_backtrace
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:170:18
>    4: rust_begin_unwind
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
>    5: core::panicking::panic_fmt
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
>    6: core::result::unwrap_failed
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
>    7: <core::result::Result<ash::instance::Instance, ash::vk::enums::Result>>::unwrap
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
>    8: <blade_graphics::hal::Context>::init_impl
>              at /home/williamdes/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:281:13
>    9: <blade_graphics::hal::Context>::init_windowed::<gpui::platform::linux::window::RawWindow>
>              at /home/williamdes/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:562:9
>   10: <gpui::platform::linux::window::LinuxWindowState>::new
>              at crates/gpui/src/platform/linux/window.rs:196:17
>   11: <gpui::platform::linux::platform::LinuxPlatform as gpui::platform::Platform>::open_window
>              at crates/gpui/src/platform/linux/platform.rs:193:35
>   12: <gpui::window::Window>::new
>              at crates/gpui/src/window.rs:332:31
>   13: <gpui::app::AppContext>::open_window::<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>::{closure#0}
>              at crates/gpui/src/app.rs:482:30
>   14: <gpui::app::AppContext>::update::<gpui::window::WindowHandle<workspace::Workspace>, <gpui::app::AppContext>::open_window<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>::{closure#0}>
>              at crates/gpui/src/app.rs:357:22
>   15: <gpui::app::AppContext>::open_window::<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>
>              at crates/gpui/src/app.rs:479:9
>   16: <gpui::app::async_context::AsyncAppContext>::open_window::<workspace::Workspace, <workspace::Workspace>::new_local::{closure#0}::{closure#0}::{closure#5}>
>              at crates/gpui/src/app/async_context.rs:133:12
>   17: <workspace::Workspace>::new_local::{closure#0}::{closure#0}
>              at crates/workspace/src/workspace.rs:838:17
>   18: <core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>>> as core::future::future::Future>::poll
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/future/future.rs:125:9
>   19: <<async_task::runnable::Builder<_>>::spawn_local::Checked<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>>>> as core::future::future::Future>::poll
>              at /home/williamdes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/runnable.rs:455:26
>   20: <async_task::raw::RawTask<<async_task::runnable::Builder<_>>::spawn_local::Checked<core::pin::Pin<alloc::boxed::Box<dyn core::future::future::Future<Output = core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>>>>, core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>, <gpui::executor::ForegroundExecutor>::spawn::inner<core::result::Result<(gpui::window::WindowHandle<workspace::Workspace>, alloc::vec::Vec<core::option::Option<core::result::Result<alloc::boxed::Box<dyn workspace::item::ItemHandle>, anyhow::Error>>>), anyhow::Error>>::{closure#0}, ()>>::run
>              at /home/williamdes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/raw.rs:557:17
>   21: <async_task::runnable::Runnable>::run
>              at /home/williamdes/.cargo/registry/src/index.crates.io-6f17d22bba15001f/async-task-4.7.0/src/runnable.rs:781:18
>   22: <gpui::platform::linux::platform::LinuxPlatform as gpui::platform::Platform>::run
>              at crates/gpui/src/platform/linux/platform.rs:144:17
>   23: <gpui::app::App>::run::<Zed::main::{closure#3}>
>              at crates/gpui/src/app.rs:138:9
>   24: Zed::main
>              at crates/zed/src/main.rs:124:5
>   25: <fn() as core::ops::function::FnOnce<()>>::call_once
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
>   26: std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:154:18
>   27: std::rt::lang_start::<()>::{closure#0}
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:167:18
>   28: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284:13
>       std::panicking::try::do_call
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
>       std::panicking::try
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
>       std::panic::catch_unwind
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
>       std::rt::lang_start_internal::{{closure}}
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:48
>       std::panicking::try::do_call
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
>       std::panicking::try
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
>       std::panic::catch_unwind
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
>       std::rt::lang_start_internal
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:20
>   29: std::rt::lang_start::<()>
>              at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:166:17
>   30: main
>   31: <unknown>
>   32: __libc_start_main
>   33: _start
> ```

I believe this is mitigated by running in release mode `cargo run --release`. This appears to be a vulkan feature/layer not supported..?

aminya commented:
After https://github.com/kvark/zed/pull/2, I also get a white screen for a second. This is the error I get then resulting in a crash:

<details>
<summary>Details</summary>

```
> ./target/debug/Zed
VUID-VkSwapchainCreateInfoKHR-imageExtent-01274(ERROR / SPEC): msgNum: 2094043421 - Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x564c8fd0e4e0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (3840,3240), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (3840,2076), minImageExtent = (3840,2076), maxImageExtent = (3840,2076). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
    Objects: 1
        [0] 0x564c8fd0e4e0, type: 3, name: NULL
[2024-02-05T10:47:19-08:00 ERROR client::telemetry] Failed to send events: HTTP 403
Thread "main" panicked with "called `Result::unwrap()` on an `Err` value: ERROR_SURFACE_LOST_KHR" at /home/aminya/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:717:18
   0: Zed::init_panic_hook::{closure#0}
             at crates/zed/src/main.rs:583:29
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:783:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:657:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:170:18
   4: rust_begin_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:645:5
   5: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/panicking.rs:72:14
   6: core::result::unwrap_failed
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1653:5
   7: <core::result::Result<(u32, bool), ash::vk::enums::Result>>::unwrap
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/result.rs:1077:23
   8: <blade_graphics::hal::Context>::acquire_frame
             at /home/aminya/.cargo/git/checkouts/blade-b2bcd1de1cf7ab6a/62eb18d/blade-graphics/src/vulkan/init.rs:714:13
   9: <gpui::platform::linux::blade_renderer::BladeRenderer>::draw
             at crates/gpui/src/platform/linux/blade_renderer.rs:359:21
  10: <gpui::platform::linux::window::LinuxWindow as gpui::platform::PlatformWindow>::draw
             at crates/gpui/src/platform/linux/window.rs:397:9
  11: <gpui::window::WindowContext>::draw
             at crates/gpui/src/window.rs:1074:9
  12: <gpui::window::Window>::new::{closure#0}::{closure#0}::{closure#0}
             at crates/gpui/src/window.rs:345:52
  13: <gpui::app::AppContext as gpui::Context>::update_window::<(), <gpui::window::Window>::new::{closure#0}::{closure#0}::{closure#0}>::{closure#0}
             at crates/gpui/src/app.rs:1225:26
  14: <gpui::app::AppContext>::update::<core::result::Result<(), anyhow::Error>, <gpui::app::AppContext as gpui::Context>::update_window<(), <gpui::window::Window>::new::{closure#0}::{closure#0}::{closure#0}>::{closure#0}>
             at crates/gpui/src/app.rs:357:22
  15: <gpui::app::AppContext as gpui::Context>::update_window::<(), <gpui::window::Window>::new::{closure#0}::{closure#0}::{closure#0}>
             at crates/gpui/src/app.rs:1216:9
  16: <gpui::app::async_context::AsyncAppContext as gpui::Context>::update_window::<(), <gpui::window::Window>::new::{closure#0}::{closure#0}::{closure#0}>
             at crates/gpui/src/app/async_context.rs:70:9
  17: <gpui::window::AnyWindowHandle>::update::<gpui::app::async_context::AsyncAppContext, (), <gpui::window::Window>::new::{closure#0}::{closure#0}::{closure#0}>
             at crates/gpui/src/window.rs:2561:9
  18: <gpui::window::Window>::new::{closure#0}::{closure#0}
             at crates/gpui/src/window.rs:345:21
  19: util::measure::<(), <gpui::window::Window>::new::{closure#0}::{closure#0}>
             at crates/util/src/util.rs:175:9
  20: <gpui::window::Window>::new::{closure#0}
             at crates/gpui/src/window.rs:344:17
  21: <alloc::boxed::Box<dyn core::ops::function::FnMut<(), Output = ()>> as core::ops::function::FnMut<()>>::call_mut
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2014:9
  22: <gpui::platform::linux::window::LinuxWindowState>::expose
             at crates/gpui/src/platform/linux/window.rs:246:13
  23: <gpui::platform::linux::platform::LinuxPlatform as gpui::platform::Platform>::run
             at crates/gpui/src/platform/linux/platform.rs:121:21
  24: <gpui::app::App>::run::<Zed::main::{closure#3}>
             at crates/gpui/src/app.rs:138:9
  25: Zed::main
             at crates/zed/src/main.rs:124:5
  26: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:250:5
  27: std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/sys_common/backtrace.rs:154:18
  28: std::rt::lang_start::<()>::{closure#0}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:167:18
  29: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/core/src/ops/function.rs:284:13
      std::panicking::try::do_call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
      std::panicking::try
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
      std::panic::catch_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal::{{closure}}
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:48
      std::panicking::try::do_call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:552:40
      std::panicking::try
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panicking.rs:516:19
      std::panic::catch_unwind
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/panic.rs:142:14
      std::rt::lang_start_internal
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:148:20
  30: std::rt::lang_start::<()>
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/std/src/rt.rs:166:17
  31: main
  32: __libc_start_call_main
             at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
  33: __libc_start_main_impl
             at ./csu/../csu/libc-start.c:392:3
  34: _start
```

</details>

williamdes commented:
Thank you, after https://github.com/kvark/zed/pull/2/commits/88dfb35346c370257b33faa9dbf15160aeb08ef2 that installed the missing dependency.

I now have this crash:

<details><summary>trace</summary>
<p>

```
VUID-VkSwapchainCreateInfoKHR-imageExtent-01274(ERROR / SPEC): msgNum: 2094043421 - Validation Error: [ VUID-VkSwapchainCreateInfoKHR-imageExtent-01274 ] Object 0: handle = 0x5650fc2070c0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x7cd0911d | vkCreateSwapchainKHR() called with imageExtent = (1920,1080), which is outside the bounds returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(): currentExtent = (1920,1024), minImageExtent = (1920,1024), maxImageExtent = (1920,1024). The Vulkan spec states: imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface (https://www.khronos.org/registry/vulkan/specs/1.3-khr-extensions/html/vkspec.html#VUID-VkSwapchainCreateInfoKHR-imageExtent-01274)
    Objects: 1
        [0] 0x5650fc2070c0, type: 3, name: NULL
Thread "main" panicked with "called `Result::unwrap()` on an `Err` value: glyph not found for character 'm'" at crates/editor/src/element.rs:1907:18
   0: Zed::init_panic_hook::{closure#0}
             at crates/zed/src/main.rs:583:29
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library/alloc/src/boxed.rs:2021:9
      std::panicking::rust_panic_with_hook

```

</p>
</details>

kvark committed:
Refactor cli and gpui dependnecies based on PR review feedback

kvark committed:
linux: introduce platform callbacks

mikayla-maki committed:
add a few more libraries to the linux script

kvark committed:
blade: cull mask support for sprites

kvark commented:
@williamdes yes, that's the big boss of the text system missing. Looks like the suggested path to it is implementing `cosmic_text`.

williamdes commented:
I do not know where https://github.com/zed-industries/zed/pull/7343/commits/7509677003da3398d0df796eca9c5435a12f576e comes from but:
- libxcb-dev should be libxcb1-dev
- alsa-base was removed in oldstable: https://tracker.debian.org/pkg/alsa-base see https://bugs.debian.org/852455 reading the [changelog](https://tracker.debian.org/media/packages/m/mkchromecast/changelog-0.3.9git20200902db2964a-2.1) of the package mentioned it was replaced by kmod

/cc @mikayla-maki 
The packages in the commit do not exist on Debian 12

h3mosphere commented:
@aminya: Can you try the following branch: https://github.com/h3mosphere/zed/tree/resize-on-draw

I had the same issue, this is probably a performance killer until @kvark can work out a better solution. But it works for me.

> After [kvark#2](https://github.com/kvark/zed/pull/2), I also get a white screen for a second. This is the error I get then resulting in a crash:
>

aminya commented:
> I do not know where https://github.com/zed-industries/zed/commit/7509677003da3398d0df796eca9c5435a12f576e comes from but:

It should be reverted cause it also causes conflicts for https://github.com/kvark/zed/pull/2. @mikayla-maki was this commit intended? If so, could you create pull requests instead of directly pushing?

mikayla-maki commented:
@aminya, sorry about that meant to push to a branch. Thanks for fixing things :)

noverby commented:
@bhavya3024 No, there is still missing essential functionality. E.g. font rendering.

aminya committed:
Fixes for file-watching, user assets, and system dependencies (#2)

* fix: avoid panics in case of non-existing path for watching

* fix: copy the themes and plugins

* Revert "add a few more libraries to the linux script"

This reverts commit 7509677003da3398d0df796eca9c5435a12f576e.

* fix: add vulkan validation layers to the system deps

* fix: fix the themes paths

kvark committed:
linux: query window geometry for determining the surface extents

bhavya3024 commented:
@noverby ok, waiting for the day when I can have this as an alternative to vscode.

mikayla-maki committed:
Mark TODOs and prep for merging main

mikayla-maki committed:
Merge branch 'main' into kvark-linux

mikayla-maki committed:
Fix a few out of date warnings

mikayla-maki committed:
Restore nanoid dependency

mikayla-maki committed:
Merge branch 'main' into kvark-linux

RustoMCSpit commented:
> Zed team wants to use off-the-shelf components from Rust ecosystem, then Blade is certainly at disadvantage here, since it's not widely used. It would rely on Zed team adding necessary features in a branch, then maybe upstreaming some of them.

they could also just offer its developer a full-time, work-at-home job as theyre more proficient with linux and they could then gear blade into its direction rather than other way round. just throwing the idea out there
